### ISSUE TITLE
How to visualize the modified counterfactual attention and the original attention?

### ISSUE THREAD

**Issue by natsunoshion** [The-Martyr/CausalMM#2]
Interesting work! I wonder if you could provide some codes for visualizing attention (as in the case in the appendix of the paper)?

**Comment by The-Martyr** [comment]
In a class that defines layer attention, you can use the following function directly on attn_output in the **forward** function:

'''
def visualize_attention(self, attn_weights, save_path):
        import matplotlib.pyplot as plt
        import seaborn as sns

        # weight2numpy -> map
        #attn_weights_np = attn_weights.to(float32)
        attn_weights_np = attn_weights.detach().cpu().numpy()
        plt.figure(figsize=(10, 8))
        #sns.heatmap(attn_weights_np[0, 0], cmap='viridis')  # sample1 & head 1
        sns.heatmap(attn_weights_np[0, 0], cmap='coolwarm', vmin=0, vmax=0.1, cbar=True)
        plt.title('Attention Weights Heatmap')
        plt.xlabel('Key Positions')
        plt.ylabel('Query Positions')
        plt.savefig(save_path) 
        plt.close()  
'''

Hope this helps😊
